pkg/retry/unconfirmed_message_handler.go

```diff
@@ -0,0 +1,116 @@
+package retry
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var (
+	log                 = logging.LoggerForModule()
+	defaultBaseInterval = 1 * time.Minute
+)
+
+// UnconfirmedMessageHandlerImpl informs the caller whether a resending should happen based on receiving ACK messsages.
+// Assumption: Time to receive an ACK is generally much shorter than the interval between sending consequtive messages.
+type UnconfirmedMessageHandlerImpl struct {
+	// baseInterval defines the delay after which we resend a message
+	baseInterval time.Duration
+	// resendInterval is a multiply of the baseInterval and defines how much time should pass before the next retry
+	resendInterval time.Duration
+	// ticker controls the delay after which we resend a message
+	ticker *time.Ticker
+	// ch will produce a message when the message should be resent
+	ch chan struct{}
+	// retry counts the number of retries for a given message
+	retry int
+	// retryMux is a mutex for retry variable
+	retryMux *sync.Mutex
```

Comment:
`retry` could be an atomic integer

Reply:
I tried to change `s.retry++` to `atomic.AddInt32(&s.retry, int32(1))` and removing the mutex, but this causes race to detected.

Reply:
After the recent changes, the atomic + retry actually does the job. The problem is now with `if s.numUnackedSendings > 1 {`